README
sfeed_curses
------------

sfeed_curses is a curses UI front-end for sfeed.

It shows the TAB-separated feed items in a graphical command-line UI. The
interface has a look inspired by the mutt mail client. It has a sidebar panel
for the feeds, a panel with a listing of the items and a small statusbar for
the selected item/URL. Some functions like searching and scrolling are
integrated in the interface itself.


Build and install
-----------------

$ make
# make install


Usage
-----

Like the format programs included in sfeed you can run it like this:

	sfeed_curses ~/.sfeed/feeds/*

... or by reading from stdin:

	sfeed_curses < ~/.sfeed/feeds/xkcd

By default sfeed_curses marks the items of the last day as new/bold. To manage
read/unread items in a different way a plain-text file with a list of the read
URLs can be used. To enable this behaviour the path to this file can be
specified by setting the environment variable $SFEED_URL_FILE to the URL file:

	SFEED_URL_FILE=~/.sfeed/urls sfeed_curses ~/.sfeed/feeds/*

There is a shellscript "sfeed_markread" to process the read and unread items.
See the man page for more detailed information.


Dependencies
------------

- C compiler (C99).
- libc (recommended: C99 and POSIX >= 200809).
- curses (typically ncurses), optional but recommended: but see minicurses.h.


Optional dependencies
---------------------

- POSIX make(1) for Makefile.
- mandoc for documentation: https://mdocml.bsd.lv/


Run-time dependencies
---------------------

- A (POSIX) shell.
- A terminal (emulator) supporting UTF-8 and the used capabilities.


Optional run-time dependencies
------------------------------

- xclip for yanking the URL or enclosure. See $SFEED_YANKER to change it.
- xdg-open, used as a plumber by default. See $SFEED_PLUMBER to change it.
- awk, used by the sfeed_content and sfeed_markread script.
  See the ENVIRONMENT VARIABLES section in the man page to change it.
- lynx, used by the sfeed_content script to convert HTML content.
  See the ENVIRONMENT VARIABLES section in the man page to change it.


OS tested
---------

- Linux (compilers: clang, gcc, tcc, libc: glibc, musl).
- OpenBSD (clang, gcc).
- NetBSD
- FreeBSD
- DragonFlyBSD
- Illumos (OpenIndiana).
- Windows (cygwin gcc + mintty).
- HaikuOS


Known terminal issues
---------------------

Below lists some bugs or missing features in terminals that are found while
testing.  Some of them might be fixed already upstream:

- cygwin + mintty: the xterm mouse-encoding of the mouse position is broken for
  scrolling.
- HaikuOS terminal: the xterm mouse-encoding of the mouse button number of the
  middle-button, right-button is incorrect / reversed.
- putty: the full reset attribute (ESC c, typically `rs1`) does not reset the
  window title.


Color themes
------------

To change the default theme you can set SFEED_THEME using make or in the
Makefile or include the a header file in sfeed_curses.c. See also the themes/
directory.


Running custom commands inside the program
------------------------------------------

Running commands inside the program can be useful for example to sync items or
mark all items across all feeds as read. It can be comfortable to have a
keybind for this inside the program to perform a scripted action and then
reload the feeds by sending the signal SIGHUP.

In the input handling code you can then add a case:

	case 'M':
		forkexec((char *[]) { "markallread.sh", NULL }, 0);
		break;

or

	case 'S':
		forkexec((char *[]) { "syncnews.sh", NULL }, 1);
		break;

The specified script should be in $PATH or an absolute path.

Example of a `markallread.sh` shellscript to mark all URLs as read:

	#!/bin/sh
	# mark all items/URLs as read.

	tmp=$(mktemp)
	(cat ~/.sfeed/urls; cut -f 3 ~/.sfeed/feeds/*) | \
	awk '!x[$0]++' > "$tmp" &&
	mv "$tmp" ~/.sfeed/urls &&
	pkill -SIGHUP sfeed_curses # reload feeds.

Example of a `syncnews.sh` shellscript to update the feeds and reload them:

	#!/bin/sh
	sfeed_update && pkill -SIGHUP sfeed_curses


Open an URL directly in the same terminal
-----------------------------------------

To open an URL directly in the same terminal using the text-mode lynx browser:

	SFEED_PLUMBER=lynx SFEED_PLUMBER_INTERACTIVE=1 sfeed_curses ~/.sfeed/feeds/*


Yank to tmux buffer
-------------------

This changes the yank command to set the tmux buffer, instead of X11 xclip:

	SFEED_YANKER="tmux set-buffer \`cat\`"


License
-------

ISC, see LICENSE file.


Author
------

Hiltjo Posthuma <hiltjo@codemadness.org>


debian/changelog
sfeed-curses (1.0+leste-1) unstable; urgency=medium

  * Merge branch 'master' into maemo/beowulf

 -- Ivan J. <parazyd@dyne.org>  Wed, 18 Aug 2021 12:38:11 +0200

sfeed-curses (0.9.10+leste-1) unstable; urgency=medium

  * Merge branch 'master' into maemo/beowulf

 -- Ivan J. <parazyd@dyne.org>  Tue, 13 Apr 2021 08:38:23 +0200

sfeed-curses (0.9.8+leste-1) unstable; urgency=medium

  * Initial Leste packaging.

 -- Ivan J. <parazyd@dyne.org>  Tue, 02 Feb 2021 13:47:48 +0100


minicurses.h
#include <sys/ioctl.h>

#undef  OK
#define OK  (0)

const char *clr_eol = "\x1b[K";
const char *clear_screen = "\x1b[H\x1b[2J";
const char *cursor_address = "\x1b[%d;%dH";
const char *cursor_normal = "\x1b[?25h"; /* DECTCEM (in)Visible cursor */
const char *cursor_invisible = "\x1b[?25l"; /* DECTCEM (in)Visible cursor */
const char *eat_newline_glitch = (void *)1;
const char *enter_ca_mode = "\x1b[?1049h"; /* smcup */
const char *exit_ca_mode = "\x1b[?1049l"; /* rmcup */
const char *save_cursor = "\x1b""7";
const char *restore_cursor = "\x1b""8";
const char *exit_attribute_mode = "\x1b[0m";
const char *enter_bold_mode = "\x1b[1m";
const char *enter_dim_mode = "\x1b[2m";
const char *enter_reverse_mode = "\x1b[7m";

int
setupterm(char *term, int fildes, int *errret)
{
	return OK;
}

char *
tparm(const char *s, int p1, int p2, ...)
{
	static char buf[32];

	if (s == cursor_address) {
		snprintf(buf, sizeof(buf), s, p1 + 1, p2 + 1);
		return buf;
	}

	return (char *)s;
}


themes/mono.h
/* default mono theme */
#define THEME_ITEM_NORMAL()           do {                            } while(0)
#define THEME_ITEM_FOCUS()            do {                            } while(0)
#define THEME_ITEM_BOLD()             do { attrmode(ATTR_BOLD_ON);    } while(0)
#define THEME_ITEM_SELECTED()         do { if (p->focused) attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_SCROLLBAR_FOCUS()       do {                            } while(0)
#define THEME_SCROLLBAR_NORMAL()      do { attrmode(ATTR_FAINT_ON);   } while(0)
#define THEME_SCROLLBAR_TICK_FOCUS()  do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_SCROLLBAR_TICK_NORMAL() do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_STATUSBAR()             do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_INPUT_LABEL()           do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_INPUT_NORMAL()          do {                            } while(0)


content.example.sh
#!/bin/sh
# RSS/Atom content viewer.
#
# Dependencies:
# - awk, sh, etc.
# - lynx or webdump or HTML converting to plain-text.

# content()
content() {
	awk -F '\t' '
function unescape(s) {
	gsub("\\\\t", "\t", s);
	gsub("\\\\n", "\n", s);
	gsub("\\\\\\\\", "\\", s);
	return s;
}
{
	print unescape($4);
	exit;
}'
}

tmp=$(mktemp 'sfeed_curses_XXXXXXXXXX')
trap "rm $tmp" EXIT
cat > "$tmp"

(awk -F '\t' '
{
	print "Title:     " $2;
	if (length($7))
		print "Author:    " $7;
	if (length($3))
		print "Link:      " $3;
	if (length($8))
		print "Enclosure: " $8;
	print "";
	exit;
}' < "$tmp"

contenttype=$(awk -F '\t' '{ print $5; exit }' < "$tmp")
if test x"$contenttype" = x"plain"; then
	content < "$tmp"
else
	(echo "<span>"; content < "$tmp";echo "</span>") | \
		lynx -stdin -dump -underline_links -image_links
#		lynx -stdin -dump -underline_links -list_inline -image_links
#		webdump -a -l -r -w 79
fi
) | \
less -R


themes/mono_highlight.h
/* mono theme with highlighting of the active panel.
   The faint attribute may not work on all terminals though.
   The combination bold with faint generally does not work either. */
#define THEME_ITEM_NORMAL()           do {                            } while(0)
#define THEME_ITEM_FOCUS()            do {                            } while(0)
#define THEME_ITEM_BOLD()             do { if (!selected) attrmode(ATTR_BOLD_ON); } while(0)
#define THEME_ITEM_SELECTED()         do { if (!p->focused) attrmode(ATTR_FAINT_ON); attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_SCROLLBAR_FOCUS()       do {                            } while(0)
#define THEME_SCROLLBAR_NORMAL()      do { attrmode(ATTR_FAINT_ON);   } while(0)
#define THEME_SCROLLBAR_TICK_FOCUS()  do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_SCROLLBAR_TICK_NORMAL() do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_LINEBAR()               do { attrmode(ATTR_FAINT_ON);   } while(0)
#define THEME_STATUSBAR()             do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_INPUT_LABEL()           do { attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_INPUT_NORMAL()          do {                            } while(0)


themes/templeos.h
/* TempleOS-like (for fun and god) */
/* set true-color foreground / background, Terry would've preferred ANSI */
#define SETFGCOLOR(r,g,b)    ttywritef("\x1b[38;2;%d;%d;%dm", r, g, b)
#define SETBGCOLOR(r,g,b)    ttywritef("\x1b[48;2;%d;%d;%dm", r, g, b)

#define THEME_ITEM_NORMAL()           do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_ITEM_FOCUS()            do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_ITEM_BOLD()             do { attrmode(ATTR_BOLD_ON); SETFGCOLOR(0xaa, 0x00, 0x00); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_ITEM_SELECTED()         do { if (p->focused) attrmode(ATTR_REVERSE_ON); } while(0)
#define THEME_SCROLLBAR_FOCUS()       do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_SCROLLBAR_NORMAL()      do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_SCROLLBAR_TICK_FOCUS()  do { SETBGCOLOR(0x00, 0x00, 0xaa); SETFGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_SCROLLBAR_TICK_NORMAL() do { SETBGCOLOR(0x00, 0x00, 0xaa); SETFGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_LINEBAR()               do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_STATUSBAR()             do { ttywrite("\x1b[6m"); SETBGCOLOR(0x00, 0x00, 0xaa); SETFGCOLOR(0xff, 0xff, 0xff); } while(0) /* blink statusbar */
#define THEME_INPUT_LABEL()           do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)
#define THEME_INPUT_NORMAL()          do { SETFGCOLOR(0x00, 0x00, 0xaa); SETBGCOLOR(0xff, 0xff, 0xff); } while(0)

#undef SCROLLBAR_SYMBOL_BAR
#define SCROLLBAR_SYMBOL_BAR  "\xe2\x95\x91" /* symbol: "double vertical" */
#undef LINEBAR_SYMBOL_BAR
#define LINEBAR_SYMBOL_BAR     "\xe2\x95\x90" /* symbol: "double horizontal" */
#undef LINEBAR_SYMBOL_RIGHT
#define LINEBAR_SYMBOL_RIGHT   "\xe2\x95\xa3" /* symbol: "double vertical and left" */
